Gems/Multiplayer/Code/Tests/NetworkEntityTests.cpp

```diff
@@ -364,6 +365,122 @@ namespace Multiplayer
+        EXPECT_TRUE(m_entityReplicationManager->HandleEntityUpdateMessage(m_mockConnection.get(), header, deleteMessage));
+    }
+
+    TEST_F(MultiplayerNetworkEntityTests, EntityReplicatorDeleteMessageIncludesUpdatedProperties)
+    {
+        // When sending a delete message, the message should also contain any properties that were changed
+        // since the previous replication.
+
+        // Always claim that every packet sent was acknowledged.
+        ON_CALL(*m_mockConnection, WasPacketAcked).WillByDefault(::testing::Return(true));
+
+        // First "send" a creation message.
+        EXPECT_TRUE(m_root->m_replicator->HasChangesToPublish());
+        EXPECT_TRUE(m_root->m_replicator->PrepareToGenerateUpdatePacket());
+        const NetworkEntityUpdateMessage createMessage = m_root->m_replicator->GenerateUpdatePacket();
+        m_root->m_replicator->RecordSentPacketId(AzNetworking::PacketId{ 1 });
+        EXPECT_FALSE(createMessage.GetIsDelete());
+        
+
+        // It should be seen as sent, so nothing more to publish.
+        EXPECT_FALSE(m_root->m_replicator->HasChangesToPublish());
+
+        // Change the translation on the entity and notify that it has been dirtied.
+        // Make sure the replicator sees it as a new change.
+        AZ::TransformBus::Event(m_root->m_entity->GetId(), &AZ::TransformBus::Events::SetWorldTranslation, AZ::Vector3(1.0f, 2.0f, 3.0f));
+        m_networkEntityManager->NotifyEntitiesDirtied();
+        EXPECT_TRUE(m_root->m_replicator->HasChangesToPublish());
+
+        // Next, mark the entity as deleted.
+        m_root->m_replicator->MarkForRemoval();
+
+        // The delete should be seen as a change that needs to be sent.
+        EXPECT_TRUE(m_root->m_replicator->HasChangesToPublish());
+
+        // Generate the delete packet.
+        EXPECT_TRUE(m_root->m_replicator->PrepareToGenerateUpdatePacket());
+        const NetworkEntityUpdateMessage deleteMessage = m_root->m_replicator->GenerateUpdatePacket();
+        m_root->m_replicator->RecordSentPacketId(AzNetworking::PacketId{ 2 });
+
+        // The message should be a delete message that has a payload larger than 3 bytes.
+        // A 3-byte payload is just a header, more than 3 bytes includes property data changes.
+        EXPECT_TRUE(deleteMessage.GetIsDelete());
+        EXPECT_GT(deleteMessage.GetData()->GetSize(), 3);
+
+        // The delete should now be seen as sent too.
+        EXPECT_FALSE(m_root->m_replicator->HasChangesToPublish());
+    }
+
+    TEST_F(MultiplayerNetworkEntityTests, EntityReplicatorDeleteMessageResentUntilAcknowledged)
+    {
+        // When sending a delete message, the message should keep getting resent until it has been acknowledged.
+
+        // Start by mocking that no packets were acknowledged.
+        ON_CALL(*m_mockConnection, WasPacketAcked).WillByDefault(::testing::Return(false));
```

Comment:
This is a cool trick with gtest mocks.. I'll have to remember it.

Reply:
Heh, to me this is the whole point to having mocks, we can easily inject behaviors on a per-test basis!